src/unix/os390-syscalls.c

```diff
@@ -0,0 +1,291 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+
+#include "os390-syscalls.h"
+#include <errno.h>
+#include <stdlib.h>
+#include <assert.h>
+#include <search.h>
+
+#define CW_CONDVAR 32
+
+#pragma linkage(BPX4CTW, OS)
+#pragma linkage(BPX1CTW, OS)
+
+static int number_of_epolls;
+static struct epoll_list* global_epoll_list[MAX_EPOLL_INSTANCES];
+static uv_mutex_t global_epoll_lock;
+static uv_once_t once = UV_ONCE_INIT;
+
+int scandir(const char *maindir, struct dirent ***namelist,
+            int (*filter)(const struct dirent *),
+            int (*compar)(const struct dirent **,
+            const struct dirent **)) {
+  struct dirent **nl;
+  struct dirent *dirent;
+  unsigned count;
+  size_t allocated;
+  DIR *mdir;
+
+  nl = NULL;
+  count = 0;
+  allocated = 0;
+  mdir = opendir(maindir);
+  if (!mdir)
+    return -1;
+
+  while (1) {
+    dirent = readdir(mdir);
+    if (!dirent)
+      break;
+    if (!filter || filter(dirent)) {
+      struct dirent *copy;
+      copy = uv__malloc(sizeof(*copy));
+      if (!copy) {
+        while (count) {
+          dirent = nl[--count];
+          uv__free(dirent);
+        }
+        uv__free(nl);
+        closedir(mdir);
+        errno = ENOMEM;
+        return -1;
+      }
+      memcpy(copy, dirent, sizeof(*copy));
+
+      nl = uv__realloc(nl, count+1);
+      nl[count++] = copy;
+    }
+  }
+
+  qsort(nl, count, sizeof(struct dirent *),
+       (int (*)(const void *, const void *))compar);
+
+  closedir(mdir);
+
+  *namelist = nl;
+  return count;
+}
+
+static int isfdequal(const struct pollfd* first,
+                     const struct pollfd* second) {
+  return first->fd == second->fd ? 0 : 1;
+}
+
+static struct pollfd* findpfd(int epfd, int fd, int events) {
+  struct epoll_list *lst;
+  struct pollfd pfd;
+  struct pollfd* found;
+
+  pfd.fd = fd;
+  pfd.events = events;
+  lst = global_epoll_list[epfd];
+  if (events == 0)
+    return lfind(&pfd, &lst->items[0], &lst->size,
+                  sizeof(pfd), isfdequal);
+  else
+    return lsearch(&pfd, &lst->items[0], &lst->size,
+                  sizeof(pfd), isfdequal);
+}
+
+static void epoll_init() {
+  uv_mutex_init(&global_epoll_lock);
+}
+  
+int epoll_create1(int flags) {
+  struct epoll_list* list;
+  int index;
+
+  uv_once(&once, epoll_init);
+  uv_mutex_lock(&global_epoll_lock);
+  list = uv__malloc(sizeof(*list));
+  index = number_of_epolls++;
+  global_epoll_list[index] = list;
+  uv_mutex_unlock(&global_epoll_lock);
+
+  /* initialize list */
+  memset(list, 0, sizeof(*list));
+  if (uv_mutex_init(&list->lock)) {
```

Comment:
why do we need a lock for each epoll? Since the loop is single threaded I think we can do away without it, and simplify some code, can't we?